batch/batch/server/server.py

```diff
@@ -764,36 +813,54 @@ def update_job_with_pod(job, pod):
         job.mark_unscheduled()
 
 
-@routes.post('/pod_changed')
-async def pod_changed(request):
-    parameters = await request.json()
+@routes.get('/recent')
+@aiohttp_jinja2.template('recent.html')
+@authenticated_users_only
+async def recent(request):  # pylint: disable=W0613
+    recent_events = get_recent_events()
+    return {'recent': list(reversed(recent_events))}
 
-    pod_name = parameters['pod_name']
 
-    job = pod_name_job.get(pod_name)
+def blocking_to_async(f, *args, **kwargs):
+    return asyncio.get_event_loop().run_in_executor(
+        app['blocking_pool'], lambda: f(*args, **kwargs))
 
-    if job and not job.is_complete():
-        try:
-            pod = v1.read_namespaced_pod(
-                pod_name,
-                POD_NAMESPACE,
-                _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
-        except kube.client.rest.ApiException as exc:
-            if exc.status == 404:
-                pod = None
-            else:
-                raise
 
+class DeblockedIterator:
+    def __init__(self, it):
+        self.it = it
+
+    def __aiter__(self):
+        return self
+
+    def __anext__(self):
+        return blocking_to_async(self.it.__next__)
+
+
+def pod_changed(pod):
+    job = pod_name_job.get(pod.metadata.name)
+
+    if job and not job.is_complete():
         update_job_with_pod(job, pod)
 
-    return web.Response(status=204)
+
+async def kube_event_loop():
+    stream = kube.watch.Watch().stream(
+        v1.list_namespaced_pod,
+        POD_NAMESPACE,
+        label_selector=f'app=batch-job,hail.is/batch-instance={instance_id}')
+    async for event in DeblockedIterator(stream):
+        pod = event['object']
+        asyncio.get_event_loop().call_soon_threadsafe(
+            pod_changed, pod)
 
 
-@routes.post('/refresh_k8s_state')
-async def refresh_k8s_state(request):  # pylint: disable=W0613
+async def refresh_k8s_state():  # pylint: disable=W0613
     log.info('started k8s state refresh')
 
-    pods = v1.list_namespaced_pod(
+    pods = await blocking_to_async(
+        app['blocking_pool'],
```

Comment:
It looks like blocking_to_async accepts a function, and automatically uses the `app['blocking_pool']`

```py
def blocking_to_async(f, *args, **kwargs):
     return asyncio.get_event_loop().run_in_executor(
         app['blocking_pool'], lambda: f(*args, **kwargs))
```